### PR TITLE
EOS-19642:  Disruptive SW upgrade - 1 and 3 node VM upgrade

### PR DESCRIPTION
**Pull request by vimalraghwani20**

### CODE REVIEW

File: rpms/s3/s3rpm.spec [hunk 1]
```diff
@@ -126,21 +126,37 @@ S3 server provides S3 REST API interface support for Motr object storage.
...
-fi
-if [ -f /opt/seagate/cortx/auth/resources/keystore.properties.sample ]; then
-    cp -f /opt/seagate/cortx/auth/resources/keystore.properties.sample /tmp/keystore.properties.sample.old
-fi
-if [ -f /opt/seagate/cortx/auth/resources/authserver.properties.sample ]; then
-    cp -f /opt/seagate/cortx/auth/resources/authserver.properties.sample /tmp/authserver.properties.sample.old
+if [ $1 == 1 ];then
+    echo "pre install section"
+elif [ $1 == 2 ];then
+    echo "pre upgrade section"
```
Comment (nileshgovande) [inline rpms/s3/s3rpm.spec:136]: Pre

Comment (nileshgovande) [reply rpms/s3/s3rpm.spec:136]: Everywhere

Comment (vimalraghwani20) [reply rpms/s3/s3rpm.spec]: fixed

File: rpms/s3/s3rpm.spec [hunk 2]
```diff
@@ -428,8 +474,30 @@ rm -rf %{buildroot}
...
+    [ -f /opt/seagate/cortx/s3/s3backgrounddelete/s3_cluster.yaml ] ||
+        cp /opt/seagate/cortx/s3/s3backgrounddelete/s3_cluster.yaml.sample /opt/seagate/cortx/s3/s3backgrounddelete/s3_cluster.yaml
+    [ -f /opt/seagate/cortx/auth/resources/authserver.properties ] ||
+        cp /opt/seagate/cortx/auth/resources/authserver.properties.sample /opt/seagate/cortx/auth/resources/authserver.properties
+    [ -f /opt/seagate/cortx/auth/resources/keystore.properties ] ||
+        cp /opt/seagate/cortx/auth/resources/keystore.properties.sample /opt/seagate/cortx/auth/resources/keystore.properties
+    echo "S3 RPM Post Install section completed"
+elif [ $1 == 2 ];then
+    echo "S3 RPM Post Upgrade section started"
+    python3.6 /opt/seagate/cortx/s3/bin/merge.py
```
[the remark this answers is not in the document]
Comment (vimalraghwani20) [reply rpms/s3/s3rpm.spec]: fixed

File: rpms/s3/s3rpm.spec [hunk 3]
```diff
@@ -438,3 +506,27 @@ if [ "$openssl_version" != "1.0.2k" ] && [ "$openssl_version" != "1.1.1" ]; then
...
+    echo "S3 RPM Post Uninstall section started"
+    # remove config files.
+    rm -f /opt/seagate/cortx/s3/conf/s3config.yaml*
+    rm -f /opt/seagate/cortx/s3/s3backgrounddelete/config.yaml*
+    rm -f /opt/seagate/cortx/s3/s3backgrounddelete/s3_cluster.yaml*
+    rm -f /opt/seagate/cortx/auth/resources/authserver.properties*
+    rm -f /opt/seagate/cortx/auth/resources/keystore.properties*
+    echo "removed all S3 config files"
+    echo "S3 RPM Post Uninstall Upgrade section completed"
+fi
```
[the remark this answers is not in the document]
Comment (vimalraghwani20) [reply rpms/s3/s3rpm.spec]: fixed

File: s3backgrounddelete/s3backgrounddelete/config/s3_cluster_unsafe_attributes.yaml [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For any questions about this software or licensing,
+# please email opensource@seagate.com or cortx-questions@seagate.com.
+#
+#
+
+cluster_config:
+   cluster_id: "dummy"
```
[the remark this answers is not in the document]
Comment (vimalraghwani20) [reply s3backgrounddelete/s3backgrounddelete/config/s3_cluster_unsafe_attributes.yaml]: fixed

File: scripts/provisioning/s3_setup [hunk 5]
```diff
@@ -101,6 +104,11 @@ def main():
       PrepareCmd(args.config).process()
       sys.stdout.write('INFO: prepare successful.\n')
 
+    elif args.command == 'upgrade':
+      from upgradecmd import UpgradeCmd
```
[the remark this answers is not in the document]
Comment (vimalraghwani20) [reply scripts/provisioning/s3_setup:108]: Its aligned the same with other phases.

File: rpms/s3/s3rpm.spec [hunk 6]
```diff
@@ -126,21 +136,40 @@ S3 server provides S3 REST API interface support for Motr object storage.
...
+        cp -f /opt/seagate/cortx/auth/resources/authserver.properties.sample /tmp/authserver.properties.sample.old
+    fi
+    echo "S3 RPM Pre Upgrade section completed"
 fi
 
+################################
+# build section
+################################
 %build
+echo "S3 RPM Build section started"
```
Comment (nileshgovande) [inline rpms/s3/s3rpm.spec:171]: Add uniform prefix to all echo statements in 
1)RPM spec
2)Merge.sh file

File: scripts/provisioning/testcmd.py [hunk 7]
```diff
@@ -47,6 +47,7 @@ def process(self):
     sys.stdout.write(f"Processing {self.name} {self.url}\n")
     self.phase_prereqs_validate(self.name)
     self.phase_keys_validate(self.url, self.name)
+    self.validate_config_files(self.name)
```
Comment (nileshgovande) [inline scripts/provisioning/testcmd.py:50]: Not needed

File: rpms/s3/s3rpm.spec [hunk 8]
```diff
@@ -17,6 +17,16 @@
 # please email opensource@seagate.com or cortx-questions@seagate.com.
 #
 
+# Following diagram represent the scriptlets flags for the RPM install/upgrade/uninstall phases in spec files.
+# new -> new RPM 
+# old -> old RPM
+#-----------------------------------------------------------------------------------------------------------------------
```
Comment (nileshgovande) [inline rpms/s3/s3rpm.spec:23]: $1 = 1

File: scripts/swupdate/merge.py [hunk 9]
```diff
@@ -64,59 +72,69 @@ def upgrade_config(configFile:str, oldSampleFile:str, newSampleFile:str, unsafeA
         - diff the value in config and sample.old - if it is changed, skip
         - if it is not changed,  we will overwrite the value in cfg file from sample.new
         - if it does not exist in cfg file add the value from sample.new file to cfg file
+    - All the arrays in yaml are always overwritten
     """
 
     #If config file is not present then simply copy sample file.
+    # TODO this code should throw error in case config file is not present.
```
Comment (nileshgovande) [inline scripts/swupdate/merge.py:79]: Do not copy - add validation instead

File: auth/resources/authserver_unsafe_attributes.properties [hunk 10]
```diff
@@ -21,12 +21,4 @@
...
-defaultEndpoint=dummy
-httpPort=dummy
-httpsPort=dummy
-defaultHost=dummy
-dataSource=dummy
-ldapHost=dummy
-ldapPort=dummy
-ldapSSLPort=dummy
\ No newline at end of file
+ldap_const_key=dummy
```
[the remark this answers is not in the document]
Comment (vimalraghwani20) [reply auth/resources/authserver_unsafe_attributes.properties]: fixed

File: rpms/s3/s3rpm.spec (hunk 6 above)
Comment (vimalraghwani20) [reply rpms/s3/s3rpm.spec:171]: added [cortx-s3server-rpm] as prefix

File: scripts/provisioning/testcmd.py (hunk 7 above)
Comment (vimalraghwani20) [reply scripts/provisioning/testcmd.py:50]: fixed

File: rpms/s3/s3rpm.spec (hunk 8 above)
Comment (vimalraghwani20) [reply rpms/s3/s3rpm.spec:23]: fixed

File: scripts/swupdate/merge.py (hunk 9 above)
Comment (vimalraghwani20) [reply scripts/swupdate/merge.py:79]: fixed

File: rpms/s3/s3rpm.spec [hunk 11]
```diff
@@ -428,13 +475,65 @@ rm -rf %{buildroot}
...
+    rm -f /opt/seagate/cortx/s3/s3backgrounddelete/config.yaml*
+    rm -f /opt/seagate/cortx/s3/s3backgrounddelete/s3backgrounddelete_unsafe_attributes.yaml
+    rm -f /opt/seagate/cortx/s3/s3backgrounddelete/s3_cluster.yaml*
+    rm -f /opt/seagate/cortx/s3/s3backgrounddelete/s3_cluster_unsafe_attributes.yaml
+    rm -f /opt/seagate/cortx/auth/resources/authserver.properties*
+    rm -f /opt/seagate/cortx/auth/resources/authserver_unsafe_attributes.properties
+    rm -f /opt/seagate/cortx/auth/resources/keystore.properties*
+    rm -f /opt/seagate/cortx/auth/resources/keystore_unsafe_attributes.properties
+    echo "[cortx-s3server-rpm] INFO: Removed all S3 config files"
+    echo "[cortx-s3server-rpm] INFO: S3 RPM Post Uninstall Upgrade section completed"
```
Comment (nileshgovande) [inline rpms/s3/s3rpm.spec:538]: Remove word upgrade from here

File: rpms/s3/s3rpm.spec [hunk 12]
```diff
@@ -17,6 +17,16 @@
...
 
+# Following diagram represent the scriptlets flags for the RPM install/upgrade/uninstall phases in spec files.
+# new -> new RPM 
+# old -> old RPM
+#--------------------------------------------------------------------------------------------------------------------------------------
+#                Install RPM                                  Upgrade RPM                                   Uninstall RPM
+#                    |                                             |                                              |
+#                    |                          pre $1 == 2 (new)  | preun $1 == 1 (old)                          |
+#   pre $1 = 1 (new) | post $1 == 1 (new)       post $1 == 2 (new) | postun $1 == 1 (old)     preun $1 == 0 (new) | postun $1 == 0(new)
+#--------------------------------------------------------------------------------------------------------------------------------------
```
Comment (nileshgovande) [inline rpms/s3/s3rpm.spec:28]: pre $1 = 1 (new) | post $1 == 1 (new)       post $1 == 2 (new) | postun $1 == 1 (old)     preun $1 == 0 (current) | postun $1 == 0(current)

File: rpms/s3/s3rpm.spec (hunk 11 above)
Comment (vimalraghwani20) [reply rpms/s3/s3rpm.spec:538]: this is to differentiate between actual post uninstall(rpm -e) and post uninstall during upgrade(rpm -Uvh)

File: rpms/s3/s3rpm.spec (hunk 12 above)
Comment (vimalraghwani20) [reply rpms/s3/s3rpm.spec:28]: not clear.

**Comment by vimalraghwani20** [comment]
http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Mini-Provisioning/job/cortx-s3server/1061/
http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Mini-Provisioning/job/cortx-s3server/1061/

**Review by nileshgovande — APPROVED** [review]

**Review by SeagateChaDeepak — APPROVED** [review]
+1

File: rpms/s3/s3rpm.spec (hunk 2 above)
[the remark this answers is not in the document]
Comment (vimalraghwani20) [reply rpms/s3/s3rpm.spec:499]: I tried to use python 36 but it was not available on mini provisioned vm. 
hence reverted back to python 3.6

**Pull request merged**